File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public static int search(int[] nums, int target, boolean bool){
        int low = 0;
        int high = nums.length -1;
        int bound = -1;
        while(low <= high){
            int mid = (low +high)/2;
            if(nums[mid] == target){
                bound = mid;
                if(bool == true){
                    high = mid -1;
                }else{
                    low = mid+1;
                }
            }else if(nums[mid] < target){
                low = mid + 1;
            }else {
                high = mid - 1;
            }
        }
        return bound;
    } 
    public int[] searchRange(int[] nums, int target) {
        if(nums.length == 0){
            return new int[]{-1,-1};
        }
        boolean bool = true;
        int first = search(nums, target, bool);
        int last = search(nums, target, !bool);
        return new int[] {first, last};
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums:
            return [-1, -1]  
        def findBound(isFirst):
            low, high = 0, len(nums) - 1
            bound = -1    
            while low <= high:
                mid = (low + high) // 2
                if nums[mid] == target:
                    bound = mid
                    if isFirst:
                        high = mid - 1  # Search left for first occurrence
                    else:
                        low = mid + 1   # Search right for last occurrence
                elif nums[mid] < target:
                    low = mid + 1
                else:
                    high = mid - 1
            return bound       
        first = findBound(True)
        last = findBound(False)
        return [first, last]
